hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/HddsConfigKeys.java

```diff
@@ -81,15 +81,19 @@
   public static final String HDDS_SCM_SAFEMODE_PIPELINE_AVAILABILITY_CHECK =
       "hdds.scm.safemode.pipeline-availability.check";
   public static final boolean
-      HDDS_SCM_SAFEMODE_PIPELINE_AVAILABILITY_CHECK_DEFAULT = false;
+      HDDS_SCM_SAFEMODE_PIPELINE_AVAILABILITY_CHECK_DEFAULT = true;
+
+  public static final String HDDS_SCM_SAFEMODE_PIPELINE_CREATION =
```

Comment:
HDDS_SCM_SAFEMODE_PIPELINE_CREATION is there any reason for this config parameter. Because if HDDS_SCM_SAFEMODE_PIPELINE_AVAILABILITY_CHECK_DEFAULT set to true, and HDDS_SCM_SAFEMODE_PIPELINE_CREATION this is set to true, we start background pipeline creator when scm is still in safe mode, if false we start background pipeline creator after SCM is out of safe mode.

So, on a fresh cluster HDDS_SCM_SAFEMODE_PIPELINE_CREATION should be always set to true otherwise we shall never come out of safemode. So, I think it is better to run this by default(Without any config check) when pipelineAvailabilityCheck is set to true. (In this way, with less config and incorrect configuration issues handling is not required)

Reply:
Hi @bharatviswa504 , thanks for review the code.  I agree that we should not introduce unnecessary new properties.  The reason to introduce HDDS_SCM_SAFEMODE_PIPELINE_CREATION  is, 
1. HDDS_SCM_SAFEMODE_PIPELINE_AVAILABILITY_CHECK_DEFAULT controls whether HealthyPipelineSafeModeRule and OneReplicaPipelineSafeModeRule should be enabled in safemode check.  By default it's true. I only see one false in a unit test. 
2. HDDS_SCM_SAFEMODE_PIPELINE_CREATION  controls whether to start pipeline creation early in safenode.  By default it's also true. It's only used in unit test since some tests want to test pipeline creation logic after the safemode exits.  In these cases, HDDS_SCM_SAFEMODE_PIPELINE_CREATION is set to false.  
3. Some unit test, such as TestHealthyPipelineSafeModeRule,  only set HDDS_SCM_SAFEMODE_PIPELINE_AVAILABILITY_CHECK_DEFAULT to false, has no effect,  will fail the test.  In these cases, HDDS_SCM_SAFEMODE_PIPELINE_CREATION set to false is a must. 
4. It seems these two properties's real values are both in unit test scope.  As the default value of these two properties are both true, which is what we expected in production environment.  Given reason 3, let's keep both properties first.  Actutally Maybe we can try to remove HDDS_SCM_SAFEMODE_PIPELINE_AVAILABILITY_CHECK_DEFAULT, since only one false code is so far.

Reply:
Thanks for the clear explanation. From explanation looks like mostly it is required for tests. For now, we can leave the property as is, (As if we remove it might need a lot of tests to be updated, and few tests need that)

The only downside I see, if someone has set the HDDS_SCM_SAFEMODE_PIPELINE_CREATION mistakenly to false when HDDS_SCM_SAFEMODE_PIPELINE_CREATION is set to true, they never come out of safe mode on a freshly installed cluster. So, can we add some explanation in the configuration that HDDS_SCM_SAFEMODE_PIPELINE_CREATION should not be set to false, when HDDS_SCM_SAFEMODE_PIPELINE_CREATION is set to true?

Reply:
Do you mean HDDS_SCM_SAFEMODE_PIPELINE_CREATION  is set to false while HDDS_SCM_SAFEMODE_PIPELINE_AVAILABILITY_CHECK_DEFAULT is set to true?

Reply:
Yes.

Reply:
> From explanation looks like mostly it is required for tests. For now, we can leave the property as is, (As if we remove it might need a lot of tests to be updated, and few tests need that)

Can a follow-up Jira be created to remove this settings in the future? (this is a huge patch, I agree that it's better to separate out a bigger cleanup...)